### ISSUE TITLE
XGBRegressor Training with GPU - Best Approach for Training Data, X - CPU / GPU

### ISSUE THREAD

**Issue by jayanti-k-88** [dmlc/xgboost#11526]
This query is a follow up on the issue https://github.com/dmlc/xgboost/issues/9791 - Fallback to DMatrix due to mismatched devices.
I used 2 approaches: 

**1. Training Data X on CPU and model trained on GPU:** 

```
X = df_transformed.drop(columns=['label'])
y = df_transformed['label']
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42, shuffle=True)
model = XGBRegressor(tree_method='hist', device='cuda', **params)
model.fit(X_train, y_train)
y_pred = model.predict(X_val)
```

**2. Training data X on GPU and model trained on GPU as well:** 

```
X = df_transformed.drop(columns=['label'])
y = df_transformed['label']
**X = cp.array(X)** -> converted dataframe into CuPy array
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42, shuffle=True)
model = XGBRegressor(tree_method='hist', device='cuda', **params)
model.fit(X_train, y_train)
y_pred = model.predict(X_val)
```

The predictions that I got with only the CuPy change had much better R-squared value in the 1st case as compared to the second case. The first case gave me a warning about DMatrix fallback. Between the 2 cases, the **only change was the CuPy array assignment** to X - with no other change in the logic / parameters.

I went through different ways in which DMatrix is internally created in XGBoost in the following link: 

https://xgboost.readthedocs.io/en/latest/dev/group__DMatrix.html

But it is not clear, as to why the CPU case is giving better metrics in my experiment.
Could someone please explain ?
Also, what is the best approach while training on GPU - is it better to copy data from CPU to GPU or use methods like CuPy ?

**Comment by trivialfis** [comment]
Could you please provide a self-contained example that I can run and reproduce the issue?

**Comment by jayanti-k-88** [comment]
Please find the code below: 

```
X = df_transformed.drop(columns=['label'])
y = df_transformed['label']
X = cp.array(X) # only to be applied in second case
X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42, shuffle=True
)

def objective(trial):
    # hyperparameters
    params = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 1000),
        'max_depth': trial.suggest_int('max_depth', 3, 15),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.3, log=True),
        'subsample': trial.suggest_float('subsample', 0.6, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.6, 1.0),
        "gamma": trial.suggest_float("gamma", 0, 5),
        'reg_alpha': trial.suggest_float('reg_alpha', 0, 5),
        'reg_lambda': trial.suggest_float('reg_lambda', 0, 5),
        'random_state': 42,
        'n_jobs': -1
    }

    # Model
    model = XGBRegressor(tree_method='hist', device='cuda', **params)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_val)
    rmse = np.sqrt(mean_squared_error(y_val, y_pred))

    return rmse

study = optuna.create_study(direction="minimize")
study.optimize(objective, n_trials=50)  # Run 50 trials

print("Best RMSE:", study.best_value)
print("Best hyperparameters:", study.best_params)

# Train final model with best params on full data
final_model = XGBRegressor(tree_method='hist', device='cuda', **study.best_params)
final_model.fit(X, y)

y_pred = final_model.predict(X_val)

rmse = np.sqrt(mean_squared_error(y_val, y_pred))
print("RMSE:", rmse)
print("R^2:", r2_score(y_val, y_pred))
```

Please note: R-squared is not known for unseen data - I mentioned it by mistake. But, for the unseen data (test data, X_test), the **Pearson's correlation coefficient** is being evaluated between the _predicted 'label'_ values and the _actual 'label'_ values.
And the results are as follows: 

1st case: CPU - GPU DMatrix: **0.04769**
2nd case: GPU - GPU DMatrix: **0.03991**

Of course, my model needs to be improvised for better results, but I just had a doubt at this stage related to the above scenario.
My training data has 406612 rows and test data has 538150 rows. As the sample sizes are very large, the difference in the correlation coefficients mentioned above is statistically significant.

**Comment by trivialfis** [comment]
Hmm, I can't inspect the internal data structure without the appropriate data. With different input types (CPU vs. GPU), the quantile sketching algorithm has a small difference. If you are using the native interface, which is used by the `XGBregressor` internally, you can inspect the cut value through the `QuantileDMatrix`.

**Comment by jayanti-k-88** [comment]
@trivialfis Please find below, the Drive link to the data: 

https://drive.google.com/drive/folders/1kvup3ar_JIXWyS_M5hWqhQYH4flpSYeW

I have written and executed a shorter self-contained code for reproducing the issue: 

```
from google.colab import drive
drive.mount('/content/drive')

import numpy as np
import pandas as pd
import cupy as cp
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split

df = pd.read_parquet('/content/drive/MyDrive/drw-crypto-market-prediction/train.parquet')
# Create a boolean DataFrame indicating where inf or -inf appear
inf_mask = df.isin([np.inf, -np.inf])

# Sum of infinite values per column
inf_counts = inf_mask.sum()

# Filter to columns that actually have inf values
cols_with_inf = inf_counts[inf_counts > 0]

cols_to_drop = cols_with_inf.index.tolist()
df.drop(columns=cols_to_drop, inplace=True)

X = df.drop(columns=['label'])
#X = cp.array(X) # - to be uncommented in 2nd case
y = df['label']

X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42, shuffle=True
)

params = {
  'n_estimators': 899,
  'max_depth': 8,
  'learning_rate': 0.13596928056342733,
  'subsample': 0.6361593213270959,
  'colsample_bytree': 0.9119804821338976,
  'gamma': 0.022051350642786703,
  'reg_alpha': 2.163652684623028,
  'reg_lambda': 4.109645135195266
}

model = XGBRegressor(tree_method='hist', device='cuda', **params)
model.fit(X_train, y_train)
```

Please execute the code and let me know.
The Google Drive link can be accessed and mounted.

**Comment by trivialfis** [comment]
Yeah, the quantile sketching algorithm produced a slightly different result, then the rest is changed.

**Comment by jayanti-k-88** [comment]
@trivialfis thanks for the confirmation. I believe this needs to be changed in the library ? Would you like to brief me with the next steps ?

**Comment by trivialfis** [comment]
I don't think there's any meaningful change at the moment. It's challenging to ensure that two different implementations produce the same result, mainly due to floating-point errors and slight differences in the algorithm. There is some random variance in the outcome.

I can take a closer look to reduce the variance after I sort out the work at hand. But in general, decision tree is quite sensitive to floating point differences.

**Comment by jayanti-k-88** [comment]
@trivialfis any update on this please? Could you please provide details of what the change will be? And, also suggest the best approach for creating DMatrix, so as to get the best results.

**Comment by trivialfis** [comment]
Not yet. As mentioned, some variance is expected. This has been a headache for a while.

For a longer explanation:

Decision trees are extremely sensitive, hence known for overfitting the dataset. A small change in a single split can lead to all the following trees changing dramatically. It's not like a neural network, in which a small difference in one weight update might not matter much for the overall model weights.

At the moment, I consider this more of a "fact of life" instead of a bug. I can **try** to reduce the variance, but I wouldn't promise that it can lead to improved results. I use the term "variance" since it can lead to both good and bad results. In your case, the GPU led to a bad one. But that's not a rule.

And the bad one is a validation dataset, so in an unwelcome way, the GPU is actually more "accurate" and overfits the data.

In summary, yes, I would like to reduce the variance. But in the end, it might not meet your current expectations.

Lastly:

>  And, also suggest the best approach for creating DMatrix, so as to get the best results.

Both of your options are fine, it's mostly due to memory usage concerns when people choose the CPU path. There's no reliable way to guess which one would improve accuracy.

**Comment by trivialfis** [comment]
Closing as not actionable
